// @rnmapbox/maps/lib/commonjs/components/HeatmapLayer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.NATIVE_MODULE_NAME = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _reactNative = require("react-native");
var _utils = require("../utils");
var _styleMap = require("../utils/styleMap");
var _AbstractLayer = _interopRequireDefault(require("./AbstractLayer"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
const MapboxGL = _reactNative.NativeModules.MGLModule;
const NATIVE_MODULE_NAME = 'RCTMGLHeatmapLayer';

/**
 * HeatmapLayer is a style layer that renders one or more filled circles on the map.
 */
exports.NATIVE_MODULE_NAME = NATIVE_MODULE_NAME;
class HeatmapLayer extends _AbstractLayer.default {
  render() {
    const props = {
      ...this.baseProps,
      sourceLayerID: this.props.sourceLayerID
    };
    return /*#__PURE__*/_react.default.createElement(RCTMGLHeatmapLayer, _extends({
      ref: this.setNativeLayer
    }, props));
  }
}
_defineProperty(HeatmapLayer, "propTypes", {
  ..._utils.viewPropTypes,
  /**
   * A string that uniquely identifies the source in the style to which it is added.
   */
  id: _propTypes.default.string.isRequired,
  /**
   * The source from which to obtain the data to style.
   * If the source has not yet been added to the current style, the behavior is undefined.
   * Inferred from parent source only if the layer is a direct child to it.
   */
  sourceID: _propTypes.default.string,
  /**
   * Identifier of the layer within the source identified by the sourceID property
   * from which the receiver obtains the data to style.
   */
  sourceLayerID: _propTypes.default.string,
  /**
   * Inserts a layer above aboveLayerID.
   */
  aboveLayerID: _propTypes.default.string,
  /**
   * Inserts a layer below belowLayerID
   */
  belowLayerID: _propTypes.default.string,
  /**
   * Inserts a layer at a specified index
   */
  layerIndex: _propTypes.default.number,
  /**
   *  Filter only the features in the source layer that satisfy a condition that you define
   */
  filter: _propTypes.default.array,
  /**
   * The minimum zoom level at which the layer gets parsed and appears.
   */
  minZoomLevel: _propTypes.default.number,
  /**
   * The maximum zoom level at which the layer gets parsed and appears.
   */
  maxZoomLevel: _propTypes.default.number,
  /**
   * Customizable style attributes
   */
  style: _propTypes.default.oneOfType([_styleMap.HeatmapLayerStyleProp, _propTypes.default.arrayOf(_styleMap.HeatmapLayerStyleProp)])
});
_defineProperty(HeatmapLayer, "defaultProps", {
  sourceID: MapboxGL.StyleSource.DefaultSourceID
});
const RCTMGLHeatmapLayer = (0, _reactNative.requireNativeComponent)(NATIVE_MODULE_NAME, HeatmapLayer, {
  nativeOnly: {
    reactStyle: true
  }
});
var _default = HeatmapLayer;
exports.default = _default;
