// @rnmapbox/maps/lib/commonjs/components/LineLayer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.NATIVE_MODULE_NAME = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _AbstractLayer = _interopRequireDefault(require("./AbstractLayer"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
const MapboxGL = _reactNative.NativeModules.MGLModule;
const NATIVE_MODULE_NAME = 'RCTMGLLineLayer';
exports.NATIVE_MODULE_NAME = NATIVE_MODULE_NAME;
/**
 * LineLayer is a style layer that renders one or more stroked polylines on the map.
 */
class LineLayer extends _AbstractLayer.default {
  static defaultProps = {
    sourceID: MapboxGL.StyleSource.DefaultSourceID
  };
  render() {
    const props = {
      ...this.baseProps,
      sourceLayerID: this.props.sourceLayerID
    };
    return /*#__PURE__*/_react.default.createElement(RCTMGLLineLayer, _extends({
      ref: this.setNativeLayer
    }, props));
  }
}
const RCTMGLLineLayer = (0, _reactNative.requireNativeComponent)(NATIVE_MODULE_NAME);
var _default = LineLayer;
exports.default = _default;
