// @rnmapbox/maps/lib/commonjs/components/ShapeSource.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ShapeSource = exports.NATIVE_MODULE_NAME = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _filterUtils = require("../utils/filterUtils");
var _utils = require("../utils");
var _deprecation = require("../utils/deprecation");
var _AbstractSource = _interopRequireDefault(require("./AbstractSource"));
var _NativeBridgeComponent = _interopRequireDefault(require("./NativeBridgeComponent"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const MapboxGL = _reactNative.NativeModules.MGLModule;
const NATIVE_MODULE_NAME = 'RCTMGLShapeSource';
exports.NATIVE_MODULE_NAME = NATIVE_MODULE_NAME;
/**
 * ShapeSource is a map content source that supplies vector shapes to be shown on the map.
 * The shape may be an url or a GeoJSON object
 */
class ShapeSource extends (0, _NativeBridgeComponent.default)(_AbstractSource.default, NATIVE_MODULE_NAME) {
  static NATIVE_ASSETS_KEY = 'assets';
  static defaultProps = {
    id: MapboxGL.StyleSource.DefaultSourceID
  };
  constructor(props) {
    super(props);
  }
  _setNativeRef(nativeRef) {
    this.setNativeRef(nativeRef);
    super._runPendingNativeCommands(nativeRef);
  }

  /**
   * Returns all features from the source that match the query parameters whether the feature is currently
   * rendered on the map.
   *
   * @example
   * shapeSource.features()
   *
   * @param  {Array=} filter - an optional filter statement to filter the returned Features.
   * @return {FeatureCollection}
   */
  async features() {
    let filter = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
    const res = await this._runNativeCommand('features', this._nativeRef, (0, _filterUtils.getFilter)(filter));
    if ((0, _utils.isAndroid)()) {
      return JSON.parse(res.data);
    }
    return res.data;
  }

  /**
   * Returns the zoom needed to expand the cluster.
   *
   * @example
   * const zoom = await shapeSource.getClusterExpansionZoom(clusterId);
   *
   * @param  {Feature} feature - The feature cluster to expand.
   * @return {number}
   */
  async getClusterExpansionZoom(feature) {
    if (typeof feature === 'number') {
      console.warn('Using cluster_id is deprecated and will be removed from the future releases. Please use cluster as an argument instead.');
      const res = await this._runNativeCommand('getClusterExpansionZoomById', this._nativeRef, [feature]);
      return res.data;
    }
    const res = await this._runNativeCommand('getClusterExpansionZoom', this._nativeRef, [JSON.stringify(feature)]);
    return res.data;
  }

  /**
   * Returns the FeatureCollection from the cluster.
   *
   * @example
   * const collection = await shapeSource.getClusterLeaves(clusterId, limit, offset);
   *
   * @param  {GeoJSON.Feature} feature - The feature cluster to expand.
   * @param  {number} limit - The number of points to return.
   * @param  {number} offset - The amount of points to skip (for pagination).
   * @return {FeatureCollection}
   */
  async getClusterLeaves(feature, limit, offset) {
    if (typeof feature === 'number') {
      console.warn('Using cluster_id is deprecated and will be removed from the future releases. Please use cluster as an argument instead.');
      const res = await this._runNativeCommand('getClusterLeavesById', this._nativeRef, [feature, limit, offset]);
      if ((0, _utils.isAndroid)()) {
        return JSON.parse(res.data);
      }
      return res.data;
    }
    const res = await this._runNativeCommand('getClusterLeaves', this._nativeRef, [JSON.stringify(feature), limit, offset]);
    if ((0, _utils.isAndroid)()) {
      return JSON.parse(res.data);
    }
    return res.data;
  }

  /**
   * Returns the FeatureCollection from the cluster (on the next zoom level).
   *
   * @example
   * const collection = await shapeSource.getClusterChildren(clusterId);
   *
   * @param  {GeoJSON.Feature} feature - The feature cluster to expand.
   * @return {FeatureCollection}
   */
  async getClusterChildren(feature) {
    if (typeof feature === 'number') {
      console.warn('Using cluster_id is deprecated and will be removed from the future releases. Please use cluster as an argument instead.');
      const res = await this._runNativeCommand('getClusterChildrenById', this._nativeRef, [feature]);
      if ((0, _utils.isAndroid)()) {
        return JSON.parse(res.data);
      }
      return res.data;
    }
    const res = await this._runNativeCommand('getClusterChildren', this._nativeRef, [JSON.stringify(feature)]);
    if ((0, _utils.isAndroid)()) {
      return JSON.parse(res.data);
    }
    return res.data;
  }
  setNativeProps(props) {
    const shallowProps = Object.assign({}, props);

    // Adds support for Animated
    if (shallowProps.shape && typeof shallowProps.shape !== 'string') {
      shallowProps.shape = JSON.stringify(shallowProps.shape);
    }
    super.setNativeProps(shallowProps);
  }
  _getShape() {
    if (!this.props.shape) {
      return;
    }
    return (0, _utils.toJSONString)(this.props.shape);
  }
  onPress(event) {
    var _this$props$onPress, _this$props;
    const {
      nativeEvent: {
        payload: {
          features,
          coordinates,
          point
        }
      }
    } = event;
    let newEvent = {
      features,
      coordinates,
      point
    };
    newEvent = (0, _deprecation.copyPropertiesAsDeprecated)(event, newEvent, key => {
      console.warn(`event.${key} is deprecated on ShapeSource#onPress, please use event.features`);
    }, {
      nativeEvent: origNativeEvent => ({
        ...origNativeEvent,
        payload: features[0]
      })
    });
    (_this$props$onPress = (_this$props = this.props).onPress) === null || _this$props$onPress === void 0 ? void 0 : _this$props$onPress.call(_this$props, newEvent);
  }
  render() {
    const props = {
      id: this.props.id,
      existing: this.props.existing,
      url: this.props.url,
      shape: this._getShape(),
      hitbox: this.props.hitbox,
      hasPressListener: (0, _utils.isFunction)(this.props.onPress),
      onMapboxShapeSourcePress: this.onPress.bind(this),
      cluster: this.props.cluster ? 1 : 0,
      clusterRadius: this.props.clusterRadius,
      clusterMaxZoomLevel: this.props.clusterMaxZoomLevel,
      clusterProperties: this.props.clusterProperties,
      maxZoomLevel: this.props.maxZoomLevel,
      buffer: this.props.buffer,
      tolerance: this.props.tolerance,
      lineMetrics: this.props.lineMetrics,
      onPress: undefined,
      ref: nativeRef => this._setNativeRef(nativeRef),
      onAndroidCallback: (0, _utils.isAndroid)() ? this._onAndroidCallback : undefined
    };
    return /*#__PURE__*/_react.default.createElement(RCTMGLShapeSource, props, (0, _utils.cloneReactChildrenWithProps)(this.props.children, {
      sourceID: this.props.id
    }));
  }
}
exports.ShapeSource = ShapeSource;
const RCTMGLShapeSource = (0, _reactNative.requireNativeComponent)(NATIVE_MODULE_NAME);
