// @rnmapbox/maps/lib/commonjs/components/SymbolLayer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SymbolLayer = exports.NATIVE_MODULE_NAME = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _AbstractLayer = _interopRequireDefault(require("./AbstractLayer"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
const MapboxGL = _reactNative.NativeModules.MGLModule;
const NATIVE_MODULE_NAME = 'RCTMGLSymbolLayer';
exports.NATIVE_MODULE_NAME = NATIVE_MODULE_NAME;
const RCTMGLSymbolLayer = (0, _reactNative.requireNativeComponent)(NATIVE_MODULE_NAME);

/**
 * SymbolLayer is a style layer that renders icon and text labels at points or along lines on the map.
 */
class SymbolLayer extends _AbstractLayer.default {
  _shouldSnapshot() {
    let isSnapshot = false;
    if (_react.default.Children.count(this.baseProps.children) <= 0) {
      return isSnapshot;
    }
    _react.default.Children.forEach(this.baseProps.children, child => {
      if ((child === null || child === void 0 ? void 0 : child.type) === _reactNative.View) {
        isSnapshot = true;
      }
    });
    return isSnapshot;
  }
  render() {
    const props = {
      ...this.baseProps,
      snapshot: this._shouldSnapshot(),
      sourceLayerID: this.props.sourceLayerID
    };
    return /*#__PURE__*/_react.default.createElement(RCTMGLSymbolLayer, _extends({
      ref: this.setNativeLayer
    }, props), this.props.children);
  }
}
exports.SymbolLayer = SymbolLayer;
_defineProperty(SymbolLayer, "defaultProps", {
  sourceID: MapboxGL.StyleSource.DefaultSourceID
});
