// @rnmapbox/maps/lib/module/utils/index.js
/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { NativeModules, findNodeHandle, Platform, Image } from 'react-native';
function getAndroidManagerInstance(module) {
  const haveViewManagerConfig = NativeModules.UIManager && NativeModules.UIManager.getViewManagerConfig;
  return haveViewManagerConfig ? NativeModules.UIManager.getViewManagerConfig(module) : NativeModules.UIManager[module];
}
function getIosManagerInstance(module) {
  return NativeModules[getIOSModuleName(module)];
}
export function isAndroid() {
  return Platform.OS === 'android';
}
export function existenceChange(cur, next) {
  if (!cur && !next) {
    return false;
  }
  return !cur && next || cur && !next;
}
export function isFunction(fn) {
  return typeof fn === 'function';
}
export function isNumber(num) {
  return typeof num === 'number' && !Number.isNaN(num);
}
export function isUndefined(obj) {
  return typeof obj === 'undefined';
}
export function isString(str) {
  return typeof str === 'string';
}
export function isBoolean(bool) {
  return typeof bool === 'boolean';
}
export function isPrimitive(value) {
  return isString(value) || isNumber(value) || isBoolean(value);
}
export function runNativeCommand(module, name, nativeRef, args) {
  const handle = findNodeHandle(nativeRef);
  if (!handle) {
    throw new Error(`Could not find handle for native ref ${module}.${name}`);
  }
  const managerInstance = isAndroid() ? getAndroidManagerInstance(module) : getIosManagerInstance(module);
  if (!managerInstance) {
    throw new Error(`Could not find ${module}`);
  }
  if (isAndroid()) {
    const {
      Commands
    } = managerInstance;
    return NativeModules.UIManager.dispatchViewManagerCommand(handle, Commands._useCommandName ? name : Commands[name], args);
  }
  if (!managerInstance[name]) {
    throw new Error(`Could not find ${name} for ${module}`);
  }
  return managerInstance[name](handle, ...args);
}
export function cloneReactChildrenWithProps(children) {
  let propsToAdd = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (!children) {
    return null;
  }
  let foundChildren = null;
  if (!Array.isArray(children)) {
    foundChildren = [children];
  } else {
    foundChildren = children;
  }
  const filteredChildren = foundChildren.filter(child => !!child); // filter out falsy children, since some can be null
  return React.Children.map(filteredChildren, child => /*#__PURE__*/React.cloneElement(child, propsToAdd));
}
export function resolveImagePath(imageRef) {
  const res = Image.resolveAssetSource(imageRef);
  return res.uri;
}
export function getIOSModuleName(moduleName) {
  if (moduleName.startsWith('RCT')) {
    return moduleName.substring(3);
  }
  return moduleName;
}
export function toJSONString() {
  let json = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  return JSON.stringify(json);
}
