// @rnmapbox/maps/lib/commonjs/components/annotations/Annotation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _propTypes = _interopRequireDefault(require("prop-types"));
var _SymbolLayer = require("../SymbolLayer");
var _Animated = _interopRequireDefault(require("../../utils/animated/Animated"));
var _classes = require("../../classes");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
class Annotation extends _react.default.Component {
  constructor(props) {
    super(props);
    const shape = this._getShapeFromProps(props);
    this.state = {
      shape: props.animated ? new _classes.AnimatedPoint(shape) : shape
    };
    this.onPress = this.onPress.bind(this);
  }
  componentDidUpdate(prevProps) {
    if (!Array.isArray(this.props.coordinates)) {
      this.setState({
        shape: null
      });
      return;
    }
    const hasCoordChanged = prevProps.coordinates[0] !== this.props.coordinates[0] || prevProps.coordinates[1] !== this.props.coordinates[1];
    if (!hasCoordChanged) {
      return;
    }
    if (this.props.animated && this.state.shape) {
      // flush current animations
      this.state.shape.stopAnimation();
      this.state.shape.timing({
        coordinates: this.props.coordinates,
        easing: this.props.animationEasingFunction,
        duration: this.props.animationDuration
      }).start();
    } else if (!this.state.shape || !this.props.animated) {
      const shape = this._getShapeFromProps(this.props);
      this.setState({
        shape: this.props.animated ? new _classes.AnimatedPoint(shape) : shape
      });
    }
  }
  onPress() {
    if (this.props.onPress) {
      this.props.onPress();
    }
  }
  _getShapeFromProps() {
    let props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    const lng = props.coordinates[0] || 0;
    const lat = props.coordinates[1] || 0;
    return {
      type: 'Point',
      coordinates: [lng, lat]
    };
  }
  get symbolStyle() {
    if (!this.props.icon) {
      return undefined;
    }
    return Object.assign({}, this.props.style, {
      iconImage: this.props.icon
    });
  }
  render() {
    if (!this.props.coordinates) {
      return null;
    }
    return /*#__PURE__*/_react.default.createElement(_Animated.default.ShapeSource, {
      id: this.props.id,
      ref: "source",
      onPress: this.onPress,
      shape: this.state.shape
    }, this.symbolStyle && /*#__PURE__*/_react.default.createElement(_SymbolLayer.SymbolLayer, {
      id: `${this.props.id}-symbol`,
      style: this.symbolStyle
    }), this.props.children);
  }
}
_defineProperty(Annotation, "propTypes", {
  id: _propTypes.default.string.isRequired,
  animated: _propTypes.default.bool,
  animationDuration: _propTypes.default.number,
  animationEasingFunction: _propTypes.default.func,
  coordinates: _propTypes.default.arrayOf(_propTypes.default.number),
  onPress: _propTypes.default.func,
  children: _propTypes.default.any,
  style: _propTypes.default.any,
  icon: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number, _propTypes.default.object])
});
_defineProperty(Annotation, "defaultProps", {
  animated: false,
  animationDuration: 1000,
  animationEasingFunction: _reactNative.Easing.linear
});
var _default = Annotation;
exports.default = _default;
