// @rnmapbox/maps/lib/module/components/Callout.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
import React from 'react';
import PropTypes from 'prop-types';
import { View, Text, Animated, requireNativeComponent, StyleSheet } from 'react-native';
import { viewPropTypes } from '../utils';
export const NATIVE_MODULE_NAME = 'RCTMGLCallout';
const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
    width: 180,
    zIndex: 9999999
  },
  tip: {
    zIndex: 1000,
    marginTop: -2,
    elevation: 0,
    backgroundColor: 'transparent',
    borderTopWidth: 16,
    borderRightWidth: 8,
    borderBottomWidth: 0,
    borderLeftWidth: 8,
    borderTopColor: 'white',
    borderRightColor: 'transparent',
    borderBottomColor: 'transparent',
    borderLeftColor: 'transparent'
  },
  content: {
    position: 'relative',
    padding: 8,
    flex: 1,
    borderRadius: 3,
    borderWidth: 1,
    borderColor: 'rgba(0, 0, 0, 0.2)',
    backgroundColor: 'white'
  },
  title: {
    color: 'black',
    textAlign: 'center'
  }
});

/**
 *  Callout that displays information about a selected annotation near the annotation.
 */
class Callout extends React.PureComponent {
  get _containerStyle() {
    return [{
      position: 'absolute',
      zIndex: 999,
      backgroundColor: 'transparent'
    }, this.props.containerStyle];
  }
  get _hasChildren() {
    return React.Children.count(this.props.children) > 0;
  }
  _renderDefaultCallout() {
    return /*#__PURE__*/React.createElement(Animated.View, {
      style: [styles.container, this.props.style]
    }, /*#__PURE__*/React.createElement(View, {
      style: [styles.content, this.props.contentStyle]
    }, /*#__PURE__*/React.createElement(Text, {
      style: [styles.title, this.props.textStyle]
    }, this.props.title)), /*#__PURE__*/React.createElement(View, {
      style: [styles.tip, this.props.tipStyle]
    }));
  }
  _renderCustomCallout() {
    return /*#__PURE__*/React.createElement(Animated.View, _extends({}, this.props, {
      style: this.props.style
    }), this.props.children);
  }
  render() {
    const calloutContent = this._hasChildren ? this._renderCustomCallout() : this._renderDefaultCallout();
    return /*#__PURE__*/React.createElement(RCTMGLCallout, {
      style: this._containerStyle
    }, calloutContent);
  }
}
_defineProperty(Callout, "propTypes", {
  ...viewPropTypes,
  /**
   * String that gets displayed in the default callout.
   */
  title: PropTypes.string,
  /**
   * Style property for the Animated.View wrapper, apply animations to this
   */
  style: PropTypes.any,
  /**
   * Style property for the native RCTMGLCallout container, set at your own risk.
   */
  containerStyle: PropTypes.any,
  /**
   * Style property for the content bubble.
   */
  contentStyle: PropTypes.any,
  /**
   * Style property for the triangle tip under the content.
   */
  tipStyle: PropTypes.any,
  /**
   * Style property for the title in the content bubble.
   */
  textStyle: PropTypes.any
});
const RCTMGLCallout = requireNativeComponent(NATIVE_MODULE_NAME, Callout);
export default Callout;
