// @rnmapbox/maps/lib/module/components/Light.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
import React, { forwardRef, memo, useImperativeHandle, useRef } from 'react';
import { requireNativeComponent } from 'react-native';
import { transformStyle } from '../utils/StyleValue';
import nativeRef from '../utils/nativeRef';
export const NATIVE_MODULE_NAME = 'RCTMGLLight';
/**
 * Light represents the light source for extruded geometries
 */
function Light(props, ref) {
  const {
    style,
    ...propWithoutStyle
  } = props;
  const nativeLightRef = nativeRef(useRef(null));
  useImperativeHandle(ref, () => ({
    setNativeProps(_props) {
      var _nativeLightRef$curre;
      let propsToPass = _props;
      if (_props.style) {
        propsToPass = {
          ..._props,
          reactStyle: transformStyle(_props.style)
        };
      }
      (_nativeLightRef$curre = nativeLightRef.current) === null || _nativeLightRef$curre === void 0 ? void 0 : _nativeLightRef$curre.setNativeProps(propsToPass);
    }
  }));
  return /*#__PURE__*/React.createElement(RCTMGLLight, _extends({
    ref: nativeLightRef,
    testID: "rctmglLight"
  }, propWithoutStyle, {
    reactStyle: transformStyle(style)
  }));
}
const RCTMGLLight = requireNativeComponent(NATIVE_MODULE_NAME);
export default /*#__PURE__*/memo( /*#__PURE__*/forwardRef(Light));
