// @rnmapbox/maps/lib/module/components/PointAnnotation.js
import React from 'react';
import { requireNativeComponent, StyleSheet, Platform } from 'react-native';
import { toJSONString, isFunction } from '../utils';
import checkRequiredProps from '../utils/checkRequiredProps';
import { makePoint } from '../utils/geoUtils';
import NativeBridgeComponent from './NativeBridgeComponent';
export const NATIVE_MODULE_NAME = 'RCTMGLPointAnnotation';
const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
    position: 'absolute'
  }
});
/**
 * PointAnnotation represents a one-dimensional shape located at a single geographical coordinate.
 *
 * Consider using ShapeSource and SymbolLayer instead, if you have many points and you have static images,
 * they'll offer much better performance.
 *
 * If you need interactive views please use MarkerView,
 * as with PointAnnotation child views are rendered onto a bitmap
 */
class PointAnnotation extends NativeBridgeComponent(React.PureComponent, NATIVE_MODULE_NAME) {
  static defaultProps = {
    anchor: {
      x: 0.5,
      y: 0.5
    },
    draggable: false
  };
  _nativeRef = null;
  constructor(props) {
    super(props);
    checkRequiredProps('PointAnnotation', props, ['id', 'coordinate']);
    this._onSelected = this._onSelected.bind(this);
    this._onDeselected = this._onDeselected.bind(this);
    this._onDragStart = this._onDragStart.bind(this);
    this._onDrag = this._onDrag.bind(this);
    this._onDragEnd = this._onDragEnd.bind(this);
  }
  _onSelected(e) {
    if (isFunction(this.props.onSelected)) {
      this.props.onSelected(e.nativeEvent.payload);
    }
  }
  _onDeselected(e) {
    if (isFunction(this.props.onDeselected)) {
      this.props.onDeselected(e.nativeEvent.payload);
    }
  }
  _onDragStart(e) {
    if (isFunction(this.props.onDragStart)) {
      this.props.onDragStart(e.nativeEvent.payload);
    }
  }
  _onDrag(e) {
    if (isFunction(this.props.onDrag)) {
      this.props.onDrag(e.nativeEvent.payload);
    }
  }
  _onDragEnd(e) {
    if (isFunction(this.props.onDragEnd)) {
      this.props.onDragEnd(e.nativeEvent.payload);
    }
  }
  _getCoordinate() {
    if (!this.props.coordinate) {
      return undefined;
    }
    return toJSONString(makePoint(this.props.coordinate));
  }

  /**
   * On v10 and pre v10 android point annotation is rendered offscreen with a canvas into an image.
   * To rerender the image from the current state of the view call refresh.
   * Call this for example from Image#onLoad.
   */
  refresh() {
    if (Platform.OS === 'android') {
      this._runNativeCommand('refresh', this._nativeRef, []);
    } else {
      this._runNativeCommand('refresh', this._nativeRef, []);
    }
  }
  _setNativeRef(nativeRef) {
    this._nativeRef = nativeRef;
    super._runPendingNativeCommands(nativeRef);
  }
  render() {
    const props = {
      ...this.props,
      ref: nativeRef => this._setNativeRef(nativeRef),
      id: this.props.id,
      title: this.props.title,
      snippet: this.props.snippet,
      anchor: this.props.anchor,
      selected: this.props.selected,
      draggable: this.props.draggable,
      style: [this.props.style, styles.container],
      onMapboxPointAnnotationSelected: this._onSelected,
      onMapboxPointAnnotationDeselected: this._onDeselected,
      onMapboxPointAnnotationDragStart: this._onDragStart,
      onMapboxPointAnnotationDrag: this._onDrag,
      onMapboxPointAnnotationDragEnd: this._onDragEnd,
      coordinate: this._getCoordinate()
    };
    return /*#__PURE__*/React.createElement(RCTMGLPointAnnotation, props, this.props.children);
  }
}
const RCTMGLPointAnnotation = requireNativeComponent(NATIVE_MODULE_NAME);
export default PointAnnotation;
