// @rnmapbox/maps/lib/module/components/SymbolLayer.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
import React from 'react';
import { View, NativeModules, requireNativeComponent } from 'react-native';
import AbstractLayer from './AbstractLayer';
const MapboxGL = NativeModules.MGLModule;
export const NATIVE_MODULE_NAME = 'RCTMGLSymbolLayer';
const RCTMGLSymbolLayer = requireNativeComponent(NATIVE_MODULE_NAME);

/**
 * SymbolLayer is a style layer that renders icon and text labels at points or along lines on the map.
 */
export class SymbolLayer extends AbstractLayer {
  _shouldSnapshot() {
    let isSnapshot = false;
    if (React.Children.count(this.baseProps.children) <= 0) {
      return isSnapshot;
    }
    React.Children.forEach(this.baseProps.children, child => {
      if ((child === null || child === void 0 ? void 0 : child.type) === View) {
        isSnapshot = true;
      }
    });
    return isSnapshot;
  }
  render() {
    const props = {
      ...this.baseProps,
      snapshot: this._shouldSnapshot(),
      sourceLayerID: this.props.sourceLayerID
    };
    return /*#__PURE__*/React.createElement(RCTMGLSymbolLayer, _extends({
      ref: this.setNativeLayer
    }, props), this.props.children);
  }
}
_defineProperty(SymbolLayer, "defaultProps", {
  sourceID: MapboxGL.StyleSource.DefaultSourceID
});
