// @rnmapbox/maps/lib/commonjs/Mapbox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  AnimatedMapPoint: true,
  StyleURL: true,
  UserTrackingModes: true,
  Camera: true,
  UserTrackingMode: true,
  Atmosphere: true,
  MapView: true,
  Light: true,
  PointAnnotation: true,
  Annotation: true,
  Callout: true,
  UserLocation: true,
  UserLocationRenderMode: true,
  VectorSource: true,
  ShapeSource: true,
  RasterSource: true,
  RasterDemSource: true,
  ImageSource: true,
  Images: true,
  Image: true,
  FillLayer: true,
  FillExtrusionLayer: true,
  HeatmapLayer: true,
  LineLayer: true,
  CircleLayer: true,
  SkyLayer: true,
  SymbolLayer: true,
  RasterLayer: true,
  BackgroundLayer: true,
  Terrain: true,
  locationManager: true,
  offlineManager: true,
  OfflineCreatePackOptions: true,
  snapshotManager: true,
  MarkerView: true,
  Animated: true,
  AnimatedCoordinatesArray: true,
  AnimatedExtractCoordinateFromArray: true,
  AnimatedPoint: true,
  AnimatedRouteCoordinatesArray: true,
  AnimatedShape: true,
  Style: true,
  Logger: true,
  requestAndroidLocationPermissions: true,
  getAnnotationsLayerID: true
};
Object.defineProperty(exports, "Animated", {
  enumerable: true,
  get: function () {
    return _Animated.default;
  }
});
Object.defineProperty(exports, "AnimatedCoordinatesArray", {
  enumerable: true,
  get: function () {
    return _classes.AnimatedCoordinatesArray;
  }
});
Object.defineProperty(exports, "AnimatedExtractCoordinateFromArray", {
  enumerable: true,
  get: function () {
    return _classes.AnimatedExtractCoordinateFromArray;
  }
});
exports.AnimatedMapPoint = void 0;
Object.defineProperty(exports, "AnimatedPoint", {
  enumerable: true,
  get: function () {
    return _classes.AnimatedPoint;
  }
});
Object.defineProperty(exports, "AnimatedRouteCoordinatesArray", {
  enumerable: true,
  get: function () {
    return _classes.AnimatedRouteCoordinatesArray;
  }
});
Object.defineProperty(exports, "AnimatedShape", {
  enumerable: true,
  get: function () {
    return _classes.AnimatedShape;
  }
});
Object.defineProperty(exports, "Annotation", {
  enumerable: true,
  get: function () {
    return _Annotation.default;
  }
});
Object.defineProperty(exports, "Atmosphere", {
  enumerable: true,
  get: function () {
    return _Atmosphere.Atmosphere;
  }
});
Object.defineProperty(exports, "BackgroundLayer", {
  enumerable: true,
  get: function () {
    return _BackgroundLayer.default;
  }
});
Object.defineProperty(exports, "Callout", {
  enumerable: true,
  get: function () {
    return _Callout.default;
  }
});
Object.defineProperty(exports, "Camera", {
  enumerable: true,
  get: function () {
    return _Camera.Camera;
  }
});
Object.defineProperty(exports, "CircleLayer", {
  enumerable: true,
  get: function () {
    return _CircleLayer.default;
  }
});
Object.defineProperty(exports, "FillExtrusionLayer", {
  enumerable: true,
  get: function () {
    return _FillExtrusionLayer.default;
  }
});
Object.defineProperty(exports, "FillLayer", {
  enumerable: true,
  get: function () {
    return _FillLayer.default;
  }
});
Object.defineProperty(exports, "HeatmapLayer", {
  enumerable: true,
  get: function () {
    return _HeatmapLayer.default;
  }
});
Object.defineProperty(exports, "Image", {
  enumerable: true,
  get: function () {
    return _Image.default;
  }
});
Object.defineProperty(exports, "ImageSource", {
  enumerable: true,
  get: function () {
    return _ImageSource.default;
  }
});
Object.defineProperty(exports, "Images", {
  enumerable: true,
  get: function () {
    return _Images.default;
  }
});
Object.defineProperty(exports, "Light", {
  enumerable: true,
  get: function () {
    return _Light.default;
  }
});
Object.defineProperty(exports, "LineLayer", {
  enumerable: true,
  get: function () {
    return _LineLayer.default;
  }
});
Object.defineProperty(exports, "Logger", {
  enumerable: true,
  get: function () {
    return _Logger.default;
  }
});
Object.defineProperty(exports, "MapView", {
  enumerable: true,
  get: function () {
    return _MapView.default;
  }
});
Object.defineProperty(exports, "MarkerView", {
  enumerable: true,
  get: function () {
    return _MarkerView.default;
  }
});
Object.defineProperty(exports, "OfflineCreatePackOptions", {
  enumerable: true,
  get: function () {
    return _offlineManager.OfflineCreatePackOptions;
  }
});
Object.defineProperty(exports, "PointAnnotation", {
  enumerable: true,
  get: function () {
    return _PointAnnotation.default;
  }
});
Object.defineProperty(exports, "RasterDemSource", {
  enumerable: true,
  get: function () {
    return _RasterDemSource.default;
  }
});
Object.defineProperty(exports, "RasterLayer", {
  enumerable: true,
  get: function () {
    return _RasterLayer.default;
  }
});
Object.defineProperty(exports, "RasterSource", {
  enumerable: true,
  get: function () {
    return _RasterSource.default;
  }
});
Object.defineProperty(exports, "ShapeSource", {
  enumerable: true,
  get: function () {
    return _ShapeSource.ShapeSource;
  }
});
Object.defineProperty(exports, "SkyLayer", {
  enumerable: true,
  get: function () {
    return _SkyLayer.default;
  }
});
Object.defineProperty(exports, "Style", {
  enumerable: true,
  get: function () {
    return _Style.default;
  }
});
exports.StyleURL = void 0;
Object.defineProperty(exports, "SymbolLayer", {
  enumerable: true,
  get: function () {
    return _SymbolLayer.SymbolLayer;
  }
});
Object.defineProperty(exports, "Terrain", {
  enumerable: true,
  get: function () {
    return _Terrain.Terrain;
  }
});
Object.defineProperty(exports, "UserLocation", {
  enumerable: true,
  get: function () {
    return _UserLocation.default;
  }
});
Object.defineProperty(exports, "UserLocationRenderMode", {
  enumerable: true,
  get: function () {
    return _UserLocation.UserLocationRenderMode;
  }
});
Object.defineProperty(exports, "UserTrackingMode", {
  enumerable: true,
  get: function () {
    return _Camera.UserTrackingMode;
  }
});
exports.UserTrackingModes = void 0;
Object.defineProperty(exports, "VectorSource", {
  enumerable: true,
  get: function () {
    return _VectorSource.default;
  }
});
Object.defineProperty(exports, "getAnnotationsLayerID", {
  enumerable: true,
  get: function () {
    return _getAnnotationsLayerID.getAnnotationsLayerID;
  }
});
Object.defineProperty(exports, "locationManager", {
  enumerable: true,
  get: function () {
    return _locationManager.default;
  }
});
Object.defineProperty(exports, "offlineManager", {
  enumerable: true,
  get: function () {
    return _offlineManager.default;
  }
});
Object.defineProperty(exports, "requestAndroidLocationPermissions", {
  enumerable: true,
  get: function () {
    return _requestAndroidLocationPermissions.requestAndroidLocationPermissions;
  }
});
Object.defineProperty(exports, "snapshotManager", {
  enumerable: true,
  get: function () {
    return _snapshotManager.default;
  }
});
var _MGLModule = require("./MGLModule");
Object.keys(_MGLModule).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _MGLModule[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _MGLModule[key];
    }
  });
});
var _Camera = require("./components/Camera");
var _Atmosphere = require("./components/Atmosphere");
var _MapView = _interopRequireDefault(require("./components/MapView"));
var _Light = _interopRequireDefault(require("./components/Light"));
var _PointAnnotation = _interopRequireDefault(require("./components/PointAnnotation"));
var _Annotation = _interopRequireDefault(require("./components/Annotation"));
var _Callout = _interopRequireDefault(require("./components/Callout"));
var _UserLocation = _interopRequireWildcard(require("./components/UserLocation"));
var _VectorSource = _interopRequireDefault(require("./components/VectorSource"));
var _ShapeSource = require("./components/ShapeSource");
var _RasterSource = _interopRequireDefault(require("./components/RasterSource"));
var _RasterDemSource = _interopRequireDefault(require("./components/RasterDemSource"));
var _ImageSource = _interopRequireDefault(require("./components/ImageSource"));
var _Images = _interopRequireDefault(require("./components/Images"));
var _Image = _interopRequireDefault(require("./components/Image"));
var _FillLayer = _interopRequireDefault(require("./components/FillLayer"));
var _FillExtrusionLayer = _interopRequireDefault(require("./components/FillExtrusionLayer"));
var _HeatmapLayer = _interopRequireDefault(require("./components/HeatmapLayer"));
var _LineLayer = _interopRequireDefault(require("./components/LineLayer"));
var _CircleLayer = _interopRequireDefault(require("./components/CircleLayer"));
var _SkyLayer = _interopRequireDefault(require("./components/SkyLayer"));
var _SymbolLayer = require("./components/SymbolLayer");
var _RasterLayer = _interopRequireDefault(require("./components/RasterLayer"));
var _BackgroundLayer = _interopRequireDefault(require("./components/BackgroundLayer"));
var _Terrain = require("./components/Terrain");
var _locationManager = _interopRequireDefault(require("./modules/location/locationManager"));
var _offlineManager = _interopRequireWildcard(require("./modules/offline/offlineManager"));
var _snapshotManager = _interopRequireDefault(require("./modules/snapshot/snapshotManager"));
var _MarkerView = _interopRequireDefault(require("./components/MarkerView"));
var _Animated = _interopRequireDefault(require("./utils/animated/Animated"));
var _classes = require("./classes");
var _Style = _interopRequireDefault(require("./components/Style"));
var _Logger = _interopRequireDefault(require("./utils/Logger"));
var _requestAndroidLocationPermissions = require("./requestAndroidLocationPermissions");
var _getAnnotationsLayerID = require("./utils/getAnnotationsLayerID");
var _deprecation = require("./utils/deprecation");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/** @deprecated This will be removed in a future release. Use `AnimatedPoint` instead. */
const AnimatedMapPoint = (0, _deprecation.deprecatedClass)(_classes.AnimatedPoint, 'AnimatedMapPoint is deprecated please use AnimatedPoint');

// types:
exports.AnimatedMapPoint = AnimatedMapPoint;
let StyleURL = /*#__PURE__*/function (StyleURL) {
  StyleURL["Street"] = "mapbox://styles/mapbox/streets-v11";
  StyleURL["Dark"] = "mapbox://styles/mapbox/dark-v10";
  StyleURL["Light"] = "mapbox://styles/mapbox/light-v10";
  StyleURL["Outdoors"] = "mapbox://styles/mapbox/outdoors-v11";
  StyleURL["Satellite"] = "mapbox://styles/mapbox/satellite-v9";
  StyleURL["SatelliteStreet"] = "mapbox://styles/mapbox/satellite-streets-v11";
  StyleURL["TrafficDay"] = "mapbox://styles/mapbox/navigation-preview-day-v4";
  StyleURL["TrafficNight"] = "mapbox://styles/mapbox/navigation-preview-night-v4";
  return StyleURL;
}({});
/** @deprecated UserTrackingModes is deprecated use UserTrackingMode */
exports.StyleURL = StyleURL;
const UserTrackingModes = _Camera.UserTrackingMode;
exports.UserTrackingModes = UserTrackingModes;
