// @rnmapbox/maps/lib/commonjs/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cloneReactChildrenWithProps = cloneReactChildrenWithProps;
exports.existenceChange = existenceChange;
exports.getIOSModuleName = getIOSModuleName;
exports.isAndroid = isAndroid;
exports.isBoolean = isBoolean;
exports.isFunction = isFunction;
exports.isNumber = isNumber;
exports.isPrimitive = isPrimitive;
exports.isString = isString;
exports.isUndefined = isUndefined;
exports.resolveImagePath = resolveImagePath;
exports.runNativeCommand = runNativeCommand;
exports.toJSONString = toJSONString;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/* eslint-disable @typescript-eslint/no-explicit-any */

function getAndroidManagerInstance(module) {
  const haveViewManagerConfig = _reactNative.NativeModules.UIManager && _reactNative.NativeModules.UIManager.getViewManagerConfig;
  return haveViewManagerConfig ? _reactNative.NativeModules.UIManager.getViewManagerConfig(module) : _reactNative.NativeModules.UIManager[module];
}
function getIosManagerInstance(module) {
  return _reactNative.NativeModules[getIOSModuleName(module)];
}
function isAndroid() {
  return _reactNative.Platform.OS === 'android';
}
function existenceChange(cur, next) {
  if (!cur && !next) {
    return false;
  }
  return !cur && next || cur && !next;
}
function isFunction(fn) {
  return typeof fn === 'function';
}
function isNumber(num) {
  return typeof num === 'number' && !Number.isNaN(num);
}
function isUndefined(obj) {
  return typeof obj === 'undefined';
}
function isString(str) {
  return typeof str === 'string';
}
function isBoolean(bool) {
  return typeof bool === 'boolean';
}
function isPrimitive(value) {
  return isString(value) || isNumber(value) || isBoolean(value);
}
function runNativeCommand(module, name, nativeRef, args) {
  const handle = (0, _reactNative.findNodeHandle)(nativeRef);
  if (!handle) {
    throw new Error(`Could not find handle for native ref ${module}.${name}`);
  }
  const managerInstance = isAndroid() ? getAndroidManagerInstance(module) : getIosManagerInstance(module);
  if (!managerInstance) {
    throw new Error(`Could not find ${module}`);
  }
  if (isAndroid()) {
    const {
      Commands
    } = managerInstance;
    return _reactNative.NativeModules.UIManager.dispatchViewManagerCommand(handle, Commands._useCommandName ? name : Commands[name], args);
  }
  if (!managerInstance[name]) {
    throw new Error(`Could not find ${name} for ${module}`);
  }
  return managerInstance[name](handle, ...args);
}
function cloneReactChildrenWithProps(children) {
  let propsToAdd = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (!children) {
    return null;
  }
  let foundChildren = null;
  if (!Array.isArray(children)) {
    foundChildren = [children];
  } else {
    foundChildren = children;
  }
  const filteredChildren = foundChildren.filter(child => !!child); // filter out falsy children, since some can be null
  return _react.default.Children.map(filteredChildren, child => /*#__PURE__*/_react.default.cloneElement(child, propsToAdd));
}
function resolveImagePath(imageRef) {
  const res = _reactNative.Image.resolveAssetSource(imageRef);
  return res.uri;
}
function getIOSModuleName(moduleName) {
  if (moduleName.startsWith('RCT')) {
    return moduleName.substring(3);
  }
  return moduleName;
}
function toJSONString() {
  let json = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  return JSON.stringify(json);
}
