// @rnmapbox/maps/lib/commonjs/utils/styleMap.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TerrainLayerStyleProp = exports.SymbolLayerStyleProp = exports.StyleTypes = exports.SkyLayerStyleProp = exports.RasterLayerStyleProp = exports.LineLayerStyleProp = exports.LightLayerStyleProp = exports.HillshadeLayerStyleProp = exports.HeatmapLayerStyleProp = exports.FillLayerStyleProp = exports.FillExtrusionLayerStyleProp = exports.CircleLayerStyleProp = exports.BackgroundLayerStyleProp = exports.AtmosphereLayerStyleProp = void 0;
exports.getStyleType = getStyleType;
exports.styleExtras = void 0;
var _propTypes = _interopRequireDefault(require("prop-types"));
var _index = require("./index");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/* eslint-disable */
// DO NOT MODIFY
// THIS FILE IS AUTOGENERATED

const StyleTypes = {
  Constant: 'constant',
  Color: 'color',
  Transition: 'transition',
  Translation: 'translation',
  Function: 'function',
  Image: 'image',
  Enum: 'enum'
};
exports.StyleTypes = StyleTypes;
function getStyleType(styleProp) {
  if (!(0, _index.isAndroid)() && styleExtras[styleProp]) {
    return styleExtras[styleProp].iosType;
  }
  if (styleMap[styleProp]) {
    return styleMap[styleProp];
  }
  throw new Error(`${styleProp} is not a valid Mapbox layer style`);
}
const FillLayerStyleProp = _propTypes.default.shape({
  /**
   * Sorts features in ascending order based on this value. Features with a higher sort key will appear above features with a lower sort key.
   */
  fillSortKey: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Whether this layer is displayed.
   */
  visibility: _propTypes.default.oneOf(['visible', 'none']),
  /**
   * Whether or not the fill should be antialiased.
   */
  fillAntialias: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.array]),
  /**
   * The opacity of the entire fill layer. In contrast to the `fillColor`, this value will also affect the 1px stroke around the fill, if the stroke is used.
   */
  fillOpacity: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s fillOpacity property.
   */
  fillOpacityTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The color of the filled part of this layer. This color can be specified as `rgba` with an alpha component and the color's opacity will not affect the opacity of the 1px stroke, if it is used.
   *
   * @disabledBy fillPattern
   */
  fillColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s fillColor property.
   */
  fillColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The outline color of the fill. Matches the value of `fillColor` if unspecified.
   *
   * @disabledBy fillPattern
   */
  fillOutlineColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s fillOutlineColor property.
   */
  fillOutlineColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The geometry's offset. Values are [x, y] where negatives indicate left and up, respectively.
   */
  fillTranslate: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s fillTranslate property.
   */
  fillTranslateTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Controls the frame of reference for `fillTranslate`.
   *
   * @requires fillTranslate
   */
  fillTranslateAnchor: _propTypes.default.oneOfType([_propTypes.default.oneOf(['map', 'viewport']), _propTypes.default.array]),
  /**
   * Name of image in sprite to use for drawing image fills. For seamless patterns, image width and height must be a factor of two (2, 4, 8, ..., 512). Note that zoomDependent expressions will be evaluated only at integer zoom levels.
   */
  fillPattern: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string, _propTypes.default.array])
});
exports.FillLayerStyleProp = FillLayerStyleProp;
const LineLayerStyleProp = _propTypes.default.shape({
  /**
   * The display of line endings.
   */
  lineCap: _propTypes.default.oneOfType([_propTypes.default.oneOf(['butt', 'round', 'square']), _propTypes.default.array]),
  /**
   * The display of lines when joining.
   */
  lineJoin: _propTypes.default.oneOfType([_propTypes.default.oneOf(['bevel', 'round', 'miter']), _propTypes.default.array]),
  /**
   * Used to automatically convert miter joins to bevel joins for sharp angles.
   */
  lineMiterLimit: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Used to automatically convert round joins to miter joins for shallow angles.
   */
  lineRoundLimit: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Sorts features in ascending order based on this value. Features with a higher sort key will appear above features with a lower sort key.
   */
  lineSortKey: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Whether this layer is displayed.
   */
  visibility: _propTypes.default.oneOf(['visible', 'none']),
  /**
   * The opacity at which the line will be drawn.
   */
  lineOpacity: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s lineOpacity property.
   */
  lineOpacityTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The color with which the line will be drawn.
   *
   * @disabledBy linePattern
   */
  lineColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s lineColor property.
   */
  lineColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The geometry's offset. Values are [x, y] where negatives indicate left and up, respectively.
   */
  lineTranslate: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s lineTranslate property.
   */
  lineTranslateTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Controls the frame of reference for `lineTranslate`.
   *
   * @requires lineTranslate
   */
  lineTranslateAnchor: _propTypes.default.oneOfType([_propTypes.default.oneOf(['map', 'viewport']), _propTypes.default.array]),
  /**
   * Stroke thickness.
   */
  lineWidth: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s lineWidth property.
   */
  lineWidthTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Draws a line casing outside of a line's actual path. Value indicates the width of the inner gap.
   */
  lineGapWidth: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s lineGapWidth property.
   */
  lineGapWidthTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The line's offset. For linear features, a positive value offsets the line to the right, relative to the direction of the line, and a negative value to the left. For polygon features, a positive value results in an inset, and a negative value results in an outset.
   */
  lineOffset: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s lineOffset property.
   */
  lineOffsetTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Blur applied to the line, in pixels.
   */
  lineBlur: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s lineBlur property.
   */
  lineBlurTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Specifies the lengths of the alternating dashes and gaps that form the dash pattern. The lengths are later scaled by the line width. To convert a dash length to pixels, multiply the length by the current line width. Note that GeoJSON sources with `lineMetrics: true` specified won't render dashed lines to the expected scale. Also note that zoomDependent expressions will be evaluated only at integer zoom levels.
   *
   * @disabledBy linePattern
   */
  lineDasharray: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.array]),
  /**
   * Name of image in sprite to use for drawing image lines. For seamless patterns, image width must be a factor of two (2, 4, 8, ..., 512). Note that zoomDependent expressions will be evaluated only at integer zoom levels.
   */
  linePattern: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string, _propTypes.default.array]),
  /**
   * Defines a gradient with which to color a line feature. Can only be used with GeoJSON sources that specify `"lineMetrics": true`.
   *
   * @disabledBy linePattern
   */
  lineGradient: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The line part between [trimStart, trimEnd] will be marked as transparent to make a route vanishing effect. The line trimOff offset is based on the whole line range [0.0, 1.0].
   */
  lineTrimOffset: _propTypes.default.arrayOf(_propTypes.default.number)
});
exports.LineLayerStyleProp = LineLayerStyleProp;
const SymbolLayerStyleProp = _propTypes.default.shape({
  /**
   * Label placement relative to its geometry.
   */
  symbolPlacement: _propTypes.default.oneOfType([_propTypes.default.oneOf(['point', 'line', 'line-center']), _propTypes.default.array]),
  /**
   * Distance between two symbol anchors.
   */
  symbolSpacing: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * If true, the symbols will not cross tile edges to avoid mutual collisions. Recommended in layers that don't have enough padding in the vector tile to prevent collisions, or if it is a point symbol layer placed after a line symbol layer. When using a client that supports global collision detection, like Mapbox GL JS version 0.42.0 or greater, enabling this property is not needed to prevent clipped labels at tile boundaries.
   */
  symbolAvoidEdges: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.array]),
  /**
   * Sorts features in ascending order based on this value. Features with lower sort keys are drawn and placed first.  When `iconAllowOverlap` or `textAllowOverlap` is `false`, features with a lower sort key will have priority during placement. When `iconAllowOverlap` or `textAllowOverlap` is set to `true`, features with a higher sort key will overlap over features with a lower sort key.
   */
  symbolSortKey: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Determines whether overlapping symbols in the same layer are rendered in the order that they appear in the data source or by their yPosition relative to the viewport. To control the order and prioritization of symbols otherwise, use `symbolSortKey`.
   */
  symbolZOrder: _propTypes.default.oneOfType([_propTypes.default.oneOf(['auto', 'viewport-y', 'source']), _propTypes.default.array]),
  /**
   * If true, the icon will be visible even if it collides with other previously drawn symbols.
   *
   * @requires iconImage
   */
  iconAllowOverlap: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.array]),
  /**
   * If true, other symbols can be visible even if they collide with the icon.
   *
   * @requires iconImage
   */
  iconIgnorePlacement: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.array]),
  /**
   * If true, text will display without their corresponding icons when the icon collides with other symbols and the text does not.
   *
   * @requires iconImage, textField
   */
  iconOptional: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.array]),
  /**
   * In combination with `symbolPlacement`, determines the rotation behavior of icons.
   *
   * @requires iconImage
   */
  iconRotationAlignment: _propTypes.default.oneOfType([_propTypes.default.oneOf(['map', 'viewport', 'auto']), _propTypes.default.array]),
  /**
   * Scales the original size of the icon by the provided factor. The new pixel size of the image will be the original pixel size multiplied by `iconSize`. 1 is the original size; 3 triples the size of the image.
   *
   * @requires iconImage
   */
  iconSize: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Scales the icon to fit around the associated text.
   *
   * @requires iconImage, textField
   */
  iconTextFit: _propTypes.default.oneOfType([_propTypes.default.oneOf(['none', 'width', 'height', 'both']), _propTypes.default.array]),
  /**
   * Size of the additional area added to dimensions determined by `iconTextFit`, in clockwise order: top, right, bottom, left.
   *
   * @requires iconImage, textField
   */
  iconTextFitPadding: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.array]),
  /**
   * Name of image in sprite to use for drawing an image background.
   */
  iconImage: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string, _propTypes.default.array]),
  /**
   * Rotates the icon clockwise.
   *
   * @requires iconImage
   */
  iconRotate: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Size of the additional area around the icon bounding box used for detecting symbol collisions.
   *
   * @requires iconImage
   */
  iconPadding: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * If true, the icon may be flipped to prevent it from being rendered upsideDown.
   *
   * @requires iconImage
   */
  iconKeepUpright: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.array]),
  /**
   * Offset distance of icon from its anchor. Positive values indicate right and down, while negative values indicate left and up. Each component is multiplied by the value of `iconSize` to obtain the final offset in pixels. When combined with `iconRotate` the offset will be as if the rotated direction was up.
   *
   * @requires iconImage
   */
  iconOffset: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.array]),
  /**
   * Part of the icon placed closest to the anchor.
   *
   * @requires iconImage
   */
  iconAnchor: _propTypes.default.oneOfType([_propTypes.default.oneOf(['center', 'left', 'right', 'top', 'bottom', 'top-left', 'top-right', 'bottom-left', 'bottom-right']), _propTypes.default.array]),
  /**
   * Orientation of icon when map is pitched.
   *
   * @requires iconImage
   */
  iconPitchAlignment: _propTypes.default.oneOfType([_propTypes.default.oneOf(['map', 'viewport', 'auto']), _propTypes.default.array]),
  /**
   * Orientation of text when map is pitched.
   *
   * @requires textField
   */
  textPitchAlignment: _propTypes.default.oneOfType([_propTypes.default.oneOf(['map', 'viewport', 'auto']), _propTypes.default.array]),
  /**
   * In combination with `symbolPlacement`, determines the rotation behavior of the individual glyphs forming the text.
   *
   * @requires textField
   */
  textRotationAlignment: _propTypes.default.oneOfType([_propTypes.default.oneOf(['map', 'viewport', 'auto']), _propTypes.default.array]),
  /**
   * Value to use for a text label. If a plain `string` is provided, it will be treated as a `formatted` with default/inherited formatting options. SDF images are not supported in formatted text and will be ignored.
   */
  textField: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * Font stack to use for displaying text.
   *
   * @requires textField
   */
  textFont: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.string), _propTypes.default.array]),
  /**
   * Font size.
   *
   * @requires textField
   */
  textSize: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The maximum line width for text wrapping.
   *
   * @requires textField
   */
  textMaxWidth: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Text leading value for multiLine text.
   *
   * @requires textField
   */
  textLineHeight: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Text tracking amount.
   *
   * @requires textField
   */
  textLetterSpacing: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Text justification options.
   *
   * @requires textField
   */
  textJustify: _propTypes.default.oneOfType([_propTypes.default.oneOf(['auto', 'left', 'center', 'right']), _propTypes.default.array]),
  /**
   * Radial offset of text, in the direction of the symbol's anchor. Useful in combination with `textVariableAnchor`, which defaults to using the twoDimensional `textOffset` if present.
   *
   * @requires textField
   */
  textRadialOffset: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * To increase the chance of placing highPriority labels on the map, you can provide an array of `textAnchor` locations: the renderer will attempt to place the label at each location, in order, before moving onto the next label. Use `textJustify: auto` to choose justification based on anchor position. To apply an offset, use the `textRadialOffset` or the twoDimensional `textOffset`.
   *
   * @requires textField
   */
  textVariableAnchor: _propTypes.default.array,
  /**
   * Part of the text placed closest to the anchor.
   *
   * @requires textField
   *
   * @disabledBy textVariableAnchor
   */
  textAnchor: _propTypes.default.oneOfType([_propTypes.default.oneOf(['center', 'left', 'right', 'top', 'bottom', 'top-left', 'top-right', 'bottom-left', 'bottom-right']), _propTypes.default.array]),
  /**
   * Maximum angle change between adjacent characters.
   *
   * @requires textField
   */
  textMaxAngle: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The property allows control over a symbol's orientation. Note that the property values act as a hint, so that a symbol whose language doesn’t support the provided orientation will be laid out in its natural orientation. Example: English point symbol will be rendered horizontally even if array value contains single 'vertical' enum value. For symbol with point placement, the order of elements in an array define priority order for the placement of an orientation variant. For symbol with line placement, the default text writing mode is either ['horizontal', 'vertical'] or ['vertical', 'horizontal'], the order doesn't affect the placement.
   *
   * @requires textField
   */
  textWritingMode: _propTypes.default.array,
  /**
   * Rotates the text clockwise.
   *
   * @requires textField
   */
  textRotate: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Size of the additional area around the text bounding box used for detecting symbol collisions.
   *
   * @requires textField
   */
  textPadding: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * If true, the text may be flipped vertically to prevent it from being rendered upsideDown.
   *
   * @requires textField
   */
  textKeepUpright: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.array]),
  /**
   * Specifies how to capitalize text, similar to the CSS `textTransform` property.
   *
   * @requires textField
   */
  textTransform: _propTypes.default.oneOfType([_propTypes.default.oneOf(['none', 'uppercase', 'lowercase']), _propTypes.default.array]),
  /**
   * Offset distance of text from its anchor. Positive values indicate right and down, while negative values indicate left and up. If used with textVariableAnchor, input values will be taken as absolute values. Offsets along the x and yAxis will be applied automatically based on the anchor position.
   *
   * @requires textField
   *
   * @disabledBy textRadialOffset
   */
  textOffset: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.array]),
  /**
   * If true, the text will be visible even if it collides with other previously drawn symbols.
   *
   * @requires textField
   */
  textAllowOverlap: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.array]),
  /**
   * If true, other symbols can be visible even if they collide with the text.
   *
   * @requires textField
   */
  textIgnorePlacement: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.array]),
  /**
   * If true, icons will display without their corresponding text when the text collides with other symbols and the icon does not.
   *
   * @requires textField, iconImage
   */
  textOptional: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.array]),
  /**
   * Whether this layer is displayed.
   */
  visibility: _propTypes.default.oneOf(['visible', 'none']),
  /**
   * The opacity at which the icon will be drawn.
   *
   * @requires iconImage
   */
  iconOpacity: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s iconOpacity property.
   */
  iconOpacityTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The color of the icon. This can only be used with [SDF icons](/help/troubleshooting/usingRecolorableImagesInMapboxMaps/).
   *
   * @requires iconImage
   */
  iconColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s iconColor property.
   */
  iconColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The color of the icon's halo. Icon halos can only be used with [SDF icons](/help/troubleshooting/usingRecolorableImagesInMapboxMaps/).
   *
   * @requires iconImage
   */
  iconHaloColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s iconHaloColor property.
   */
  iconHaloColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Distance of halo to the icon outline.
   *
   * @requires iconImage
   */
  iconHaloWidth: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s iconHaloWidth property.
   */
  iconHaloWidthTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Fade out the halo towards the outside.
   *
   * @requires iconImage
   */
  iconHaloBlur: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s iconHaloBlur property.
   */
  iconHaloBlurTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Distance that the icon's anchor is moved from its original placement. Positive values indicate right and down, while negative values indicate left and up.
   *
   * @requires iconImage
   */
  iconTranslate: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s iconTranslate property.
   */
  iconTranslateTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Controls the frame of reference for `iconTranslate`.
   *
   * @requires iconImage, iconTranslate
   */
  iconTranslateAnchor: _propTypes.default.oneOfType([_propTypes.default.oneOf(['map', 'viewport']), _propTypes.default.array]),
  /**
   * The opacity at which the text will be drawn.
   *
   * @requires textField
   */
  textOpacity: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s textOpacity property.
   */
  textOpacityTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The color with which the text will be drawn.
   *
   * @requires textField
   */
  textColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s textColor property.
   */
  textColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The color of the text's halo, which helps it stand out from backgrounds.
   *
   * @requires textField
   */
  textHaloColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s textHaloColor property.
   */
  textHaloColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Distance of halo to the font outline. Max text halo width is 1/4 of the fontSize.
   *
   * @requires textField
   */
  textHaloWidth: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s textHaloWidth property.
   */
  textHaloWidthTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The halo's fadeout distance towards the outside.
   *
   * @requires textField
   */
  textHaloBlur: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s textHaloBlur property.
   */
  textHaloBlurTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Distance that the text's anchor is moved from its original placement. Positive values indicate right and down, while negative values indicate left and up.
   *
   * @requires textField
   */
  textTranslate: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s textTranslate property.
   */
  textTranslateTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Controls the frame of reference for `textTranslate`.
   *
   * @requires textField, textTranslate
   */
  textTranslateAnchor: _propTypes.default.oneOfType([_propTypes.default.oneOf(['map', 'viewport']), _propTypes.default.array])
});
exports.SymbolLayerStyleProp = SymbolLayerStyleProp;
const CircleLayerStyleProp = _propTypes.default.shape({
  /**
   * Sorts features in ascending order based on this value. Features with a higher sort key will appear above features with a lower sort key.
   */
  circleSortKey: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Whether this layer is displayed.
   */
  visibility: _propTypes.default.oneOf(['visible', 'none']),
  /**
   * Circle radius.
   */
  circleRadius: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s circleRadius property.
   */
  circleRadiusTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The fill color of the circle.
   */
  circleColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s circleColor property.
   */
  circleColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Amount to blur the circle. 1 blurs the circle such that only the centerpoint is full opacity.
   */
  circleBlur: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s circleBlur property.
   */
  circleBlurTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The opacity at which the circle will be drawn.
   */
  circleOpacity: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s circleOpacity property.
   */
  circleOpacityTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The geometry's offset. Values are [x, y] where negatives indicate left and up, respectively.
   */
  circleTranslate: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s circleTranslate property.
   */
  circleTranslateTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Controls the frame of reference for `circleTranslate`.
   *
   * @requires circleTranslate
   */
  circleTranslateAnchor: _propTypes.default.oneOfType([_propTypes.default.oneOf(['map', 'viewport']), _propTypes.default.array]),
  /**
   * Controls the scaling behavior of the circle when the map is pitched.
   */
  circlePitchScale: _propTypes.default.oneOfType([_propTypes.default.oneOf(['map', 'viewport']), _propTypes.default.array]),
  /**
   * Orientation of circle when map is pitched.
   */
  circlePitchAlignment: _propTypes.default.oneOfType([_propTypes.default.oneOf(['map', 'viewport']), _propTypes.default.array]),
  /**
   * The width of the circle's stroke. Strokes are placed outside of the `circleRadius`.
   */
  circleStrokeWidth: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s circleStrokeWidth property.
   */
  circleStrokeWidthTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The stroke color of the circle.
   */
  circleStrokeColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s circleStrokeColor property.
   */
  circleStrokeColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The opacity of the circle's stroke.
   */
  circleStrokeOpacity: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s circleStrokeOpacity property.
   */
  circleStrokeOpacityTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  })
});
exports.CircleLayerStyleProp = CircleLayerStyleProp;
const HeatmapLayerStyleProp = _propTypes.default.shape({
  /**
   * Whether this layer is displayed.
   */
  visibility: _propTypes.default.oneOf(['visible', 'none']),
  /**
   * Radius of influence of one heatmap point in pixels. Increasing the value makes the heatmap smoother, but less detailed. `queryRenderedFeatures` on heatmap layers will return points within this radius.
   */
  heatmapRadius: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s heatmapRadius property.
   */
  heatmapRadiusTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * A measure of how much an individual point contributes to the heatmap. A value of 10 would be equivalent to having 10 points of weight 1 in the same spot. Especially useful when combined with clustering.
   */
  heatmapWeight: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Similar to `heatmapWeight` but controls the intensity of the heatmap globally. Primarily used for adjusting the heatmap based on zoom level.
   */
  heatmapIntensity: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s heatmapIntensity property.
   */
  heatmapIntensityTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Defines the color of each pixel based on its density value in a heatmap.  Should be an expression that uses `["heatmapDensity"]` as input.
   */
  heatmapColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The global opacity at which the heatmap layer will be drawn.
   */
  heatmapOpacity: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s heatmapOpacity property.
   */
  heatmapOpacityTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  })
});
exports.HeatmapLayerStyleProp = HeatmapLayerStyleProp;
const FillExtrusionLayerStyleProp = _propTypes.default.shape({
  /**
   * Whether this layer is displayed.
   */
  visibility: _propTypes.default.oneOf(['visible', 'none']),
  /**
   * The opacity of the entire fill extrusion layer. This is rendered on a perLayer, not perFeature, basis, and dataDriven styling is not available.
   */
  fillExtrusionOpacity: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s fillExtrusionOpacity property.
   */
  fillExtrusionOpacityTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The base color of the extruded fill. The extrusion's surfaces will be shaded differently based on this color in combination with the root `light` settings. If this color is specified as `rgba` with an alpha component, the alpha component will be ignored; use `fillExtrusionOpacity` to set layer opacity.
   *
   * @disabledBy fillExtrusionPattern
   */
  fillExtrusionColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s fillExtrusionColor property.
   */
  fillExtrusionColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The geometry's offset. Values are [x, y] where negatives indicate left and up (on the flat plane), respectively.
   */
  fillExtrusionTranslate: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s fillExtrusionTranslate property.
   */
  fillExtrusionTranslateTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Controls the frame of reference for `fillExtrusionTranslate`.
   *
   * @requires fillExtrusionTranslate
   */
  fillExtrusionTranslateAnchor: _propTypes.default.oneOfType([_propTypes.default.oneOf(['map', 'viewport']), _propTypes.default.array]),
  /**
   * Name of image in sprite to use for drawing images on extruded fills. For seamless patterns, image width and height must be a factor of two (2, 4, 8, ..., 512). Note that zoomDependent expressions will be evaluated only at integer zoom levels.
   */
  fillExtrusionPattern: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string, _propTypes.default.array]),
  /**
   * The height with which to extrude this layer.
   */
  fillExtrusionHeight: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s fillExtrusionHeight property.
   */
  fillExtrusionHeightTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The height with which to extrude the base of this layer. Must be less than or equal to `fillExtrusionHeight`.
   *
   * @requires fillExtrusionHeight
   */
  fillExtrusionBase: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s fillExtrusionBase property.
   */
  fillExtrusionBaseTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Whether to apply a vertical gradient to the sides of a fillExtrusion layer. If true, sides will be shaded slightly darker farther down.
   */
  fillExtrusionVerticalGradient: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.array])
});
exports.FillExtrusionLayerStyleProp = FillExtrusionLayerStyleProp;
const RasterLayerStyleProp = _propTypes.default.shape({
  /**
   * Whether this layer is displayed.
   */
  visibility: _propTypes.default.oneOf(['visible', 'none']),
  /**
   * The opacity at which the image will be drawn.
   */
  rasterOpacity: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s rasterOpacity property.
   */
  rasterOpacityTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Rotates hues around the color wheel.
   */
  rasterHueRotate: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s rasterHueRotate property.
   */
  rasterHueRotateTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Increase or reduce the brightness of the image. The value is the minimum brightness.
   */
  rasterBrightnessMin: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s rasterBrightnessMin property.
   */
  rasterBrightnessMinTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Increase or reduce the brightness of the image. The value is the maximum brightness.
   */
  rasterBrightnessMax: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s rasterBrightnessMax property.
   */
  rasterBrightnessMaxTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Increase or reduce the saturation of the image.
   */
  rasterSaturation: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s rasterSaturation property.
   */
  rasterSaturationTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Increase or reduce the contrast of the image.
   */
  rasterContrast: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s rasterContrast property.
   */
  rasterContrastTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The resampling/interpolation method to use for overscaling, also known as texture magnification filter
   */
  rasterResampling: _propTypes.default.oneOfType([_propTypes.default.oneOf(['linear', 'nearest']), _propTypes.default.array]),
  /**
   * Fade duration when a new tile is added.
   */
  rasterFadeDuration: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array])
});
exports.RasterLayerStyleProp = RasterLayerStyleProp;
const HillshadeLayerStyleProp = _propTypes.default.shape({
  /**
   * Whether this layer is displayed.
   */
  visibility: _propTypes.default.oneOf(['visible', 'none']),
  /**
   * The direction of the light source used to generate the hillshading with 0 as the top of the viewport if `hillshadeIlluminationAnchor` is set to `viewport` and due north if `hillshadeIlluminationAnchor` is set to `map`.
   */
  hillshadeIlluminationDirection: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Direction of light source when map is rotated.
   */
  hillshadeIlluminationAnchor: _propTypes.default.oneOfType([_propTypes.default.oneOf(['map', 'viewport']), _propTypes.default.array]),
  /**
   * Intensity of the hillshade
   */
  hillshadeExaggeration: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s hillshadeExaggeration property.
   */
  hillshadeExaggerationTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The shading color of areas that face away from the light source.
   */
  hillshadeShadowColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s hillshadeShadowColor property.
   */
  hillshadeShadowColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The shading color of areas that faces towards the light source.
   */
  hillshadeHighlightColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s hillshadeHighlightColor property.
   */
  hillshadeHighlightColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The shading color used to accentuate rugged terrain like sharp cliffs and gorges.
   */
  hillshadeAccentColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s hillshadeAccentColor property.
   */
  hillshadeAccentColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  })
});
exports.HillshadeLayerStyleProp = HillshadeLayerStyleProp;
const BackgroundLayerStyleProp = _propTypes.default.shape({
  /**
   * Whether this layer is displayed.
   */
  visibility: _propTypes.default.oneOf(['visible', 'none']),
  /**
   * The color with which the background will be drawn.
   *
   * @disabledBy backgroundPattern
   */
  backgroundColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s backgroundColor property.
   */
  backgroundColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Name of image in sprite to use for drawing an image background. For seamless patterns, image width and height must be a factor of two (2, 4, 8, ..., 512). Note that zoomDependent expressions will be evaluated only at integer zoom levels.
   */
  backgroundPattern: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string, _propTypes.default.array]),
  /**
   * The opacity at which the background will be drawn.
   */
  backgroundOpacity: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s backgroundOpacity property.
   */
  backgroundOpacityTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  })
});
exports.BackgroundLayerStyleProp = BackgroundLayerStyleProp;
const SkyLayerStyleProp = _propTypes.default.shape({
  /**
   * Whether this layer is displayed.
   */
  visibility: _propTypes.default.oneOf(['visible', 'none']),
  /**
   * The type of the sky
   */
  skyType: _propTypes.default.oneOfType([_propTypes.default.oneOf(['gradient', 'atmosphere']), _propTypes.default.array]),
  /**
   * Position of the sun center [a azimuthal angle, p polar angle]. The azimuthal angle indicates the position of the sun relative to 0° north, where degrees proceed clockwise. The polar angle indicates the height of the sun, where 0° is directly above, at zenith, and 90° at the horizon. When this property is ommitted, the sun center is directly inherited from the light position.
   */
  skyAtmosphereSun: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.array]),
  /**
   * Intensity of the sun as a light source in the atmosphere (on a scale from 0 to a 100). Setting higher values will brighten up the sky.
   */
  skyAtmosphereSunIntensity: _propTypes.default.number,
  /**
   * Position of the gradient center [a azimuthal angle, p polar angle]. The azimuthal angle indicates the position of the gradient center relative to 0° north, where degrees proceed clockwise. The polar angle indicates the height of the gradient center, where 0° is directly above, at zenith, and 90° at the horizon.
   */
  skyGradientCenter: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.array]),
  /**
   * The angular distance (measured in degrees) from `skyGradientCenter` up to which the gradient extends. A value of 180 causes the gradient to wrap around to the opposite direction from `skyGradientCenter`.
   */
  skyGradientRadius: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * Defines a radial color gradient with which to color the sky. The color values can be interpolated with an expression using `skyRadialProgress`. The range [0, 1] for the interpolant covers a radial distance (in degrees) of [0, `skyGradientRadius`] centered at the position specified by `skyGradientCenter`.
   */
  skyGradient: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * A color applied to the atmosphere sun halo. The alpha channel describes how strongly the sun halo is represented in an atmosphere sky layer.
   */
  skyAtmosphereHaloColor: _propTypes.default.string,
  /**
   * A color used to tweak the main atmospheric scattering coefficients. Using white applies the default coefficients giving the natural blue color to the atmosphere. This color affects how heavily the corresponding wavelength is represented during scattering. The alpha channel describes the density of the atmosphere, with 1 maximum density and 0 no density.
   */
  skyAtmosphereColor: _propTypes.default.string,
  /**
   * The opacity of the entire sky layer.
   */
  skyOpacity: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s skyOpacity property.
   */
  skyOpacityTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  })
});
exports.SkyLayerStyleProp = SkyLayerStyleProp;
const LightLayerStyleProp = _propTypes.default.shape({
  /**
   * Whether extruded geometries are lit relative to the map or viewport.
   */
  anchor: _propTypes.default.oneOfType([_propTypes.default.oneOf(['map', 'viewport']), _propTypes.default.array]),
  /**
   * Position of the light source relative to lit (extruded) geometries, in [r radial coordinate, a azimuthal angle, p polar angle] where r indicates the distance from the center of the base of an object to its light, a indicates the position of the light relative to 0° (0° when `light.anchor` is set to `viewport` corresponds to the top of the viewport, or 0° when `light.anchor` is set to `map` corresponds to due north, and degrees proceed clockwise), and p indicates the height of the light (from 0°, directly above, to 180°, directly below).
   */
  position: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s position property.
   */
  positionTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Color tint for lighting extruded geometries.
   */
  color: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s color property.
   */
  colorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Intensity of lighting (on a scale from 0 to 1). Higher numbers will present as more extreme contrast.
   */
  intensity: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s intensity property.
   */
  intensityTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  })
});
exports.LightLayerStyleProp = LightLayerStyleProp;
const AtmosphereLayerStyleProp = _propTypes.default.shape({
  /**
   * The start and end distance range in which fog fades from fully transparent to fully opaque. The distance to the point at the center of the map is defined as zero, so that negative range values are closer to the camera, and positive values are farther away.
   */
  range: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s range property.
   */
  rangeTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The color of the atmosphere region immediately below the horizon and within the `range` and above the horizon and within `horizonBlend`. Using opacity is recommended only for smoothly transitioning fog on/off as anything less than 100% opacity results in more tiles loaded and drawn.
   */
  color: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s color property.
   */
  colorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The color of the atmosphere region above the horizon, `highColor` extends further above the horizon than the `color` property and its spread can be controlled with `horizonBlend`. The opacity can be set to `0` to remove the high atmosphere color contribution.
   */
  highColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s highColor property.
   */
  highColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * The color of the region above the horizon and after the end of the `horizonBlend` contribution. The opacity can be set to `0` to have a transparent background.
   */
  spaceColor: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s spaceColor property.
   */
  spaceColorTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * Horizon blend applies a smooth fade from the color of the atmosphere to the color of space. A value of zero leaves a sharp transition from atmosphere to space. Increasing the value blends the color of atmosphere into increasingly high angles of the sky.
   */
  horizonBlend: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s horizonBlend property.
   */
  horizonBlendTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  }),
  /**
   * A value controlling the star intensity where `0` will show no stars and `1` will show stars at their maximum intensity.
   */
  starIntensity: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s starIntensity property.
   */
  starIntensityTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  })
});
exports.AtmosphereLayerStyleProp = AtmosphereLayerStyleProp;
const TerrainLayerStyleProp = _propTypes.default.shape({
  /**
   * Name of a source of `raster_dem` type to be used for terrain elevation.
   */
  source: _propTypes.default.string,
  /**
   * Exaggerates the elevation of the terrain by multiplying the data from the DEM with this value.
   *
   * @requires source
   */
  exaggeration: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  /**
   * The transition affecting any changes to this layer’s exaggeration property.
   */
  exaggerationTransition: _propTypes.default.shape({
    duration: _propTypes.default.number,
    delay: _propTypes.default.number
  })
});
exports.TerrainLayerStyleProp = TerrainLayerStyleProp;
const styleMap = {
  fillSortKey: StyleTypes.Constant,
  fillAntialias: StyleTypes.Constant,
  fillOpacity: StyleTypes.Constant,
  fillOpacityTransition: StyleTypes.Transition,
  fillColor: StyleTypes.Color,
  fillColorTransition: StyleTypes.Transition,
  fillOutlineColor: StyleTypes.Color,
  fillOutlineColorTransition: StyleTypes.Transition,
  fillTranslate: StyleTypes.Translation,
  fillTranslateTransition: StyleTypes.Transition,
  fillTranslateAnchor: StyleTypes.Enum,
  fillPattern: StyleTypes.Image,
  lineCap: StyleTypes.Enum,
  lineJoin: StyleTypes.Enum,
  lineMiterLimit: StyleTypes.Constant,
  lineRoundLimit: StyleTypes.Constant,
  lineSortKey: StyleTypes.Constant,
  lineOpacity: StyleTypes.Constant,
  lineOpacityTransition: StyleTypes.Transition,
  lineColor: StyleTypes.Color,
  lineColorTransition: StyleTypes.Transition,
  lineTranslate: StyleTypes.Translation,
  lineTranslateTransition: StyleTypes.Transition,
  lineTranslateAnchor: StyleTypes.Enum,
  lineWidth: StyleTypes.Constant,
  lineWidthTransition: StyleTypes.Transition,
  lineGapWidth: StyleTypes.Constant,
  lineGapWidthTransition: StyleTypes.Transition,
  lineOffset: StyleTypes.Constant,
  lineOffsetTransition: StyleTypes.Transition,
  lineBlur: StyleTypes.Constant,
  lineBlurTransition: StyleTypes.Transition,
  lineDasharray: StyleTypes.Constant,
  linePattern: StyleTypes.Image,
  lineGradient: StyleTypes.Color,
  lineTrimOffset: StyleTypes.Constant,
  symbolPlacement: StyleTypes.Enum,
  symbolSpacing: StyleTypes.Constant,
  symbolAvoidEdges: StyleTypes.Constant,
  symbolSortKey: StyleTypes.Constant,
  symbolZOrder: StyleTypes.Enum,
  iconAllowOverlap: StyleTypes.Constant,
  iconIgnorePlacement: StyleTypes.Constant,
  iconOptional: StyleTypes.Constant,
  iconRotationAlignment: StyleTypes.Enum,
  iconSize: StyleTypes.Constant,
  iconTextFit: StyleTypes.Enum,
  iconTextFitPadding: StyleTypes.Constant,
  iconImage: StyleTypes.Image,
  iconRotate: StyleTypes.Constant,
  iconPadding: StyleTypes.Constant,
  iconKeepUpright: StyleTypes.Constant,
  iconOffset: StyleTypes.Constant,
  iconAnchor: StyleTypes.Enum,
  iconPitchAlignment: StyleTypes.Enum,
  textPitchAlignment: StyleTypes.Enum,
  textRotationAlignment: StyleTypes.Enum,
  textField: StyleTypes.Constant,
  textFont: StyleTypes.Constant,
  textSize: StyleTypes.Constant,
  textMaxWidth: StyleTypes.Constant,
  textLineHeight: StyleTypes.Constant,
  textLetterSpacing: StyleTypes.Constant,
  textJustify: StyleTypes.Enum,
  textRadialOffset: StyleTypes.Constant,
  textVariableAnchor: StyleTypes.Constant,
  textAnchor: StyleTypes.Enum,
  textMaxAngle: StyleTypes.Constant,
  textWritingMode: StyleTypes.Constant,
  textRotate: StyleTypes.Constant,
  textPadding: StyleTypes.Constant,
  textKeepUpright: StyleTypes.Constant,
  textTransform: StyleTypes.Enum,
  textOffset: StyleTypes.Constant,
  textAllowOverlap: StyleTypes.Constant,
  textIgnorePlacement: StyleTypes.Constant,
  textOptional: StyleTypes.Constant,
  iconOpacity: StyleTypes.Constant,
  iconOpacityTransition: StyleTypes.Transition,
  iconColor: StyleTypes.Color,
  iconColorTransition: StyleTypes.Transition,
  iconHaloColor: StyleTypes.Color,
  iconHaloColorTransition: StyleTypes.Transition,
  iconHaloWidth: StyleTypes.Constant,
  iconHaloWidthTransition: StyleTypes.Transition,
  iconHaloBlur: StyleTypes.Constant,
  iconHaloBlurTransition: StyleTypes.Transition,
  iconTranslate: StyleTypes.Translation,
  iconTranslateTransition: StyleTypes.Transition,
  iconTranslateAnchor: StyleTypes.Enum,
  textOpacity: StyleTypes.Constant,
  textOpacityTransition: StyleTypes.Transition,
  textColor: StyleTypes.Color,
  textColorTransition: StyleTypes.Transition,
  textHaloColor: StyleTypes.Color,
  textHaloColorTransition: StyleTypes.Transition,
  textHaloWidth: StyleTypes.Constant,
  textHaloWidthTransition: StyleTypes.Transition,
  textHaloBlur: StyleTypes.Constant,
  textHaloBlurTransition: StyleTypes.Transition,
  textTranslate: StyleTypes.Translation,
  textTranslateTransition: StyleTypes.Transition,
  textTranslateAnchor: StyleTypes.Enum,
  circleSortKey: StyleTypes.Constant,
  circleRadius: StyleTypes.Constant,
  circleRadiusTransition: StyleTypes.Transition,
  circleColor: StyleTypes.Color,
  circleColorTransition: StyleTypes.Transition,
  circleBlur: StyleTypes.Constant,
  circleBlurTransition: StyleTypes.Transition,
  circleOpacity: StyleTypes.Constant,
  circleOpacityTransition: StyleTypes.Transition,
  circleTranslate: StyleTypes.Translation,
  circleTranslateTransition: StyleTypes.Transition,
  circleTranslateAnchor: StyleTypes.Enum,
  circlePitchScale: StyleTypes.Enum,
  circlePitchAlignment: StyleTypes.Enum,
  circleStrokeWidth: StyleTypes.Constant,
  circleStrokeWidthTransition: StyleTypes.Transition,
  circleStrokeColor: StyleTypes.Color,
  circleStrokeColorTransition: StyleTypes.Transition,
  circleStrokeOpacity: StyleTypes.Constant,
  circleStrokeOpacityTransition: StyleTypes.Transition,
  heatmapRadius: StyleTypes.Constant,
  heatmapRadiusTransition: StyleTypes.Transition,
  heatmapWeight: StyleTypes.Constant,
  heatmapIntensity: StyleTypes.Constant,
  heatmapIntensityTransition: StyleTypes.Transition,
  heatmapColor: StyleTypes.Color,
  heatmapOpacity: StyleTypes.Constant,
  heatmapOpacityTransition: StyleTypes.Transition,
  fillExtrusionOpacity: StyleTypes.Constant,
  fillExtrusionOpacityTransition: StyleTypes.Transition,
  fillExtrusionColor: StyleTypes.Color,
  fillExtrusionColorTransition: StyleTypes.Transition,
  fillExtrusionTranslate: StyleTypes.Translation,
  fillExtrusionTranslateTransition: StyleTypes.Transition,
  fillExtrusionTranslateAnchor: StyleTypes.Enum,
  fillExtrusionPattern: StyleTypes.Image,
  fillExtrusionHeight: StyleTypes.Constant,
  fillExtrusionHeightTransition: StyleTypes.Transition,
  fillExtrusionBase: StyleTypes.Constant,
  fillExtrusionBaseTransition: StyleTypes.Transition,
  fillExtrusionVerticalGradient: StyleTypes.Constant,
  rasterOpacity: StyleTypes.Constant,
  rasterOpacityTransition: StyleTypes.Transition,
  rasterHueRotate: StyleTypes.Constant,
  rasterHueRotateTransition: StyleTypes.Transition,
  rasterBrightnessMin: StyleTypes.Constant,
  rasterBrightnessMinTransition: StyleTypes.Transition,
  rasterBrightnessMax: StyleTypes.Constant,
  rasterBrightnessMaxTransition: StyleTypes.Transition,
  rasterSaturation: StyleTypes.Constant,
  rasterSaturationTransition: StyleTypes.Transition,
  rasterContrast: StyleTypes.Constant,
  rasterContrastTransition: StyleTypes.Transition,
  rasterResampling: StyleTypes.Enum,
  rasterFadeDuration: StyleTypes.Constant,
  hillshadeIlluminationDirection: StyleTypes.Constant,
  hillshadeIlluminationAnchor: StyleTypes.Enum,
  hillshadeExaggeration: StyleTypes.Constant,
  hillshadeExaggerationTransition: StyleTypes.Transition,
  hillshadeShadowColor: StyleTypes.Color,
  hillshadeShadowColorTransition: StyleTypes.Transition,
  hillshadeHighlightColor: StyleTypes.Color,
  hillshadeHighlightColorTransition: StyleTypes.Transition,
  hillshadeAccentColor: StyleTypes.Color,
  hillshadeAccentColorTransition: StyleTypes.Transition,
  backgroundColor: StyleTypes.Color,
  backgroundColorTransition: StyleTypes.Transition,
  backgroundPattern: StyleTypes.Image,
  backgroundOpacity: StyleTypes.Constant,
  backgroundOpacityTransition: StyleTypes.Transition,
  skyType: StyleTypes.Enum,
  skyAtmosphereSun: StyleTypes.Constant,
  skyAtmosphereSunIntensity: StyleTypes.Constant,
  skyGradientCenter: StyleTypes.Constant,
  skyGradientRadius: StyleTypes.Constant,
  skyGradient: StyleTypes.Color,
  skyAtmosphereHaloColor: StyleTypes.Color,
  skyAtmosphereColor: StyleTypes.Color,
  skyOpacity: StyleTypes.Constant,
  skyOpacityTransition: StyleTypes.Transition,
  anchor: StyleTypes.Enum,
  position: StyleTypes.Constant,
  positionTransition: StyleTypes.Transition,
  intensity: StyleTypes.Constant,
  intensityTransition: StyleTypes.Transition,
  range: StyleTypes.Constant,
  rangeTransition: StyleTypes.Transition,
  highColor: StyleTypes.Color,
  highColorTransition: StyleTypes.Transition,
  spaceColor: StyleTypes.Color,
  spaceColorTransition: StyleTypes.Transition,
  horizonBlend: StyleTypes.Constant,
  horizonBlendTransition: StyleTypes.Transition,
  starIntensity: StyleTypes.Constant,
  starIntensityTransition: StyleTypes.Transition,
  source: StyleTypes.Constant,
  exaggeration: StyleTypes.Constant,
  exaggerationTransition: StyleTypes.Transition,
  color: StyleTypes.Color,
  colorTransition: StyleTypes.Transition,
  visibility: StyleTypes.Constant
};
const styleExtras = {
  // padding
  iconTextFitPadding: {
    iosType: 'edgeinsets'
  },
  // offsets
  iconOffset: {
    iosType: 'vector'
  },
  textOffset: {
    iosType: 'vector'
  },
  lineOffset: {
    iosType: 'vector'
  },
  // translates
  fillTranslate: {
    iosType: 'vector'
  },
  lineTranslate: {
    iosType: 'vector'
  },
  iconTranslate: {
    iosType: 'vector'
  },
  textTranslate: {
    iosType: 'vector'
  },
  circleTranslate: {
    iosType: 'vector'
  },
  fillExtrusionTranslate: {
    iosType: 'vector'
  }
};
exports.styleExtras = styleExtras;
var _default = styleMap;
exports.default = _default;
